File: tests/reboot/zod/transaction_with_same_state_id/servicer_api.ts
import { reader, transaction, writer } from "@reboot-dev/reboot-api";
import { z } from "zod/v4";

export const MainTest = {
  state: {
    data: z.string().default("").meta({ tag: 1 }),
  },
  methods: {
    transaction: transaction({
      request: z.object({
        mainTestData: z.string().meta({ tag: 1 }),
        secondaryTestData: z.array(z.string()).meta({ tag: 2 }),
      }),
      response: z.void(),
    }),
    getData: reader({
      request: z.object({}),
      response: z.object({
        mainTestData: z.string().meta({ tag: 1 }),
        secondaryTestData: z.array(z.string()).meta({ tag: 2 }),
      }),
    }),
  },
};

export const SecondaryTest = {
  state: {
    data: z
      .array(z.string())
      .default(() => [])
      .meta({ tag: 1 }),
  },
  methods: {
    writer: writer({
      request: z.object({
        data: z.array(z.string()).meta({ tag: 1 }),
      }),
      response: z.void(),
    }),
    getData: reader({
      request: z.object({}),
      response: z.object({
        data: z.array(z.string()).meta({ tag: 1 }),
      }),
    }),
  },
};

export const api = { MainTest, SecondaryTest };

File: tests/reboot/pydantic/transaction_with_same_state_id/servicer_api.py
from reboot.api import (
    API,
    Field,
    Methods,
    Model,
    Reader,
    Transaction,
    Type,
    Writer,
)
from typing import Optional


class MainTestState(Model):
    data: Optional[str] = Field(tag=1, default=None)


class SecondaryTestState(Model):
    data: Optional[list[str]] = Field(tag=1, default=None)


class TransactionRequest(Model):
    main_test_data: str = Field(tag=1)
    secondary_test_data: list[str] = Field(tag=2)


class GetDataResponse(Model):
    main_test_data: str = Field(tag=1)
    secondary_test_data: list[str] = Field(tag=2)


class WriterRequest(Model):
    data: list[str] = Field(tag=1)


class GetSecondaryDataResponse(Model):
    data: list[str] = Field(tag=1)


api = API(
    MainTest=Type(
        state=MainTestState,
        methods=Methods(
            transaction=Transaction(
                request=TransactionRequest,
                response=None,
                mcp=None,
            ),
            get_data=Reader(
                request=None,
                response=GetDataResponse,
                mcp=None,
            ),
        ),
    ),
    SecondaryTest=Type(
        state=SecondaryTestState,
        methods=Methods(
            writer=Writer(
                request=WriterRequest,
                response=None,
                mcp=None,
            ),
            get_data=Reader(
                request=None,
                response=GetSecondaryDataResponse,
                mcp=None,
            ),
        ),
    ),
)
